ycmd/completers/language_server/language_server_protocol.py

```diff
@@ -391,38 +301,36 @@ def DidChangeConfiguration( config ):
   } )
 
 
-def DidOpenTextDocument( file_state, file_types, file_contents ):
+def DidOpenTextDocument( file_state, file_types ):
   return BuildNotification( 'textDocument/didOpen', {
     'textDocument': {
       'uri': FilePathToUri( file_state.filename ),
       'languageId': '/'.join( file_types ),
       'version': file_state.version,
-      'text': file_contents
+      'text': file_state.contents
     }
   } )
 
 
-def DidChangeTextDocument( file_state, file_contents ):
+def DidChangeTextDocument( file_state, content_changes ):
   return BuildNotification( 'textDocument/didChange', {
     'textDocument': {
       'uri': FilePathToUri( file_state.filename ),
       'version': file_state.version,
     },
-    'contentChanges': [
-      { 'text': file_contents },
-    ]
+    'contentChanges': content_changes
   } )
 
 
-def DidSaveTextDocument( file_state, file_contents ):
+def DidSaveTextDocument( file_state ):
   params = {
     'textDocument': {
       'uri': FilePathToUri( file_state.filename ),
       'version': file_state.version,
     },
   }
-  if file_contents is not None:
-    params.update( { 'text': file_contents } )
+  if file_state.contents is not None:
+    params.update( { 'text': file_state.contents } )
```

Comment:
when we do this, should we set the file_state.changes to [] ?

Reply:
I'm honestly not sure what that would do, but I can test that.

Reply:
Nope. That didn't solve anything. Any reason for that change?

Reply:
I was wondering if sending the "full" text in the save notification could make the sequence out-of-sync, e.g. consider a series of events where the file is opened, 3 changes are made, then the file is saved. Because Ycmd requests are not sequenced, the events can overtake each other, e.g.

* `didOpen` (full contents state = `A`)
* `didChange` (first delta `A->B`, state = `B`)
* `didChange` (second delta `B->C`, state = `C`)
* `didSave` (full contents of state `D`, state = `D`)
* `didChange` (third delta `C->D`, state = `error (C->D is not valid when applied to last updated state D)`)


But I think that the text in didSave is not allowed to change the server file state (i.e. `didSave (text of state D, state remains C)`